compiler/rustc_lint/src/builtin.rs

```diff
@@ -730,6 +734,14 @@ fn type_implements_negative_copy_modulo_regions<'tcx>(
     tcx.infer_ctxt().build().predicate_must_hold_modulo_regions(&obligation)
 }
 
+fn def_or_variant_marked_non_exhaustive<'tcx>(tcx: TyCtxt<'tcx>, def: ty::AdtDef<'tcx>) -> bool {
+    def.variants()
+        .iter()
+        .map(|variant| variant.def_id)
+        .chain(iter::once(def.did()))
+        .any(|did| tcx.has_attr(did, sym::non_exhaustive))
```

Comment:
It should be better to use `is_variant_list_non_exhaustive`/`is_field_list_non_exhaustive` here instead of inspecting attributes manually.

Reply:
Ah, that is much nicer, thanks!